mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala

```diff
@@ -216,28 +217,77 @@ private[ml] class LogisticAggregator(
       s"is applied, multinomial loss will produce a result different from binary loss.")
   }
 
+  // dot of a empty vector (with intercept if any)
+  // With it as an offset, for a sparse vector,
+  // we only need to deal with the non-zero values in prediction.
+  @transient private lazy val binarySumOffset: Double = {
+    val localCoefficients = coefficientsArray
+    val localFeaturesAvg = bcFeaturesAvg.value
+    val localFeaturesStd = bcFeaturesStd.value
+    var sum = if (fitIntercept) localCoefficients(numFeaturesPlusIntercept - 1) else 0.0
+    var i = 0
+    while (i < localFeaturesAvg.length) {
+      if (localFeaturesStd(i) != 0.0) {
+        sum += localCoefficients(i) * (0.0 - localFeaturesAvg(i)) / localFeaturesStd(i)
+      }
+      i += 1
+    }
+    sum
+  }
+
+  // With it as an offset, for a sparse vector,
+  // we only need to deal with the non-zero values in gradient calculation.
+  @transient private lazy val binaryGradOffset: Array[Double] = {
+    bcFeaturesAvg.value.zip(bcFeaturesStd.value).map { case (avg, std) =>
+      if (std != 0) avg / std else 0.0
+    }
+  }
+
   /** Update gradient and loss using binary loss function. */
   private def binaryUpdateInPlace(features: Vector, weight: Double, label: Double): Unit = {
 
+    val localFeaturesAvg = bcFeaturesAvg.value
     val localFeaturesStd = bcFeaturesStd.value
     val localCoefficients = coefficientsArray
     val localGradientArray = gradientSumArray
+
+
+    // following equals to:
+    //    val margin = - {
+    //      var sum = 0.0
+    //      features.foreach { (index, value) =>
+    //        if (localFeaturesStd(index) != 0.0) {
+    //          sum += localCoefficients(index) *
+    //            (value - localFeaturesAvg(index)) / localFeaturesStd(index)
+    //        }
```

Comment:
add one more line to the comment:  = localCoefficients(index) * value / localFeaturesStd(index) - localCoefficients(index) * localFeaturesAvg(index) / localFeaturesStd(index), and then a comment that the second terms have been summed up above into the binarySumOffset.